rmw_fastrtps_shared_cpp/src/rmw_get_topic_info.cpp

```diff
@@ -0,0 +1,235 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <map>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "rmw/rmw.h"
+#include "rmw/types.h"
+#include "rmw/topic_info_array.h"
+
+#include "rmw_fastrtps_shared_cpp/custom_participant_info.hpp"
+#include "rmw_fastrtps_shared_cpp/rmw_common.hpp"
+#include "rmw_fastrtps_shared_cpp/namespace_prefix.hpp"
+
+namespace rmw_fastrtps_shared_cpp
+{
+
+rmw_ret_t
+_validate_params(
+  const char * identifier,
+  const rmw_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * topic_name,
+  rmw_topic_info_array_t * participants_info)
+{
+  if (!identifier) {
+    RMW_SET_ERROR_MSG("null implementation identifier provided");
+    return RMW_RET_ERROR;
+  }
+
+  if (!topic_name) {
+    RMW_SET_ERROR_MSG("null topic_name provided");
+    return RMW_RET_ERROR;
+  }
+
+  if (!allocator) {
+    RMW_SET_ERROR_MSG("null allocator provided");
+    return RMW_RET_ERROR;
+  }
+
+  if (!node) {
+    RMW_SET_ERROR_MSG("null node handle");
+    return RMW_RET_ERROR;
+  }
+
+  // Get participant pointer from node
+  if (node->implementation_identifier != identifier) {
+    RMW_SET_ERROR_MSG("node handle not from this implementation");
+    return RMW_RET_ERROR;
+  }
+
+  if (!participants_info) {
+    RMW_SET_ERROR_MSG("null participants_info provided");
+    return RMW_RET_ERROR;
+  }
+
+  return RMW_RET_OK;
+}
+
+std::vector<std::string>
+_get_topic_fqdns(const char * topic_name, bool no_mangle)
+{
+  std::vector<std::string> topic_fqdns;
+  topic_fqdns.push_back(topic_name);
+  // if mangle
+  if (!no_mangle) {
+    auto ros_prefixes = _get_all_ros_prefixes();
```

Comment:
I know this is imitating the `rmw_count` implementation, but trying with all prefixes is wrong.
I will continue the discussion in a follow up issue.

Reply:
Let's not forget this point.